11 спринт. 1 Ревью.

EvgenyAlexandrovichM opened the pull request:
Исправлен ErrorHandler.java. Добавлен перехват серверных ошибок Старый сервис переименован в storage, новый сервис реализует логику ТЗ Добавлены новые эндпоинты для новой бизнес логики.

EvgenyAlexandrovichM committed:
11 спринт. 1 Ревью.
Исправлен ErrorHandler.java. Добавлен перехват серверных ошибок
Старый сервис переименован в storage, новый сервис реализует логику ТЗ
Добавлены новые эндпоинты для новой бизнес логики.

EvgenyAlexandrovichM committed:
11 спринт. 1 Ревью.
Исправлен ErrorHandler.java. Добавлен перехват серверных ошибок
Старый сервис переименован в storage, новый сервис реализует логику ТЗ
Добавлены новые эндпоинты для новой бизнес логики.

EvgenyAlexandrovichM committed:
11 спринт. 1 Ревью.
Исправлен ErrorHandler.java. Добавлен перехват серверных ошибок
Старый сервис переименован в storage, новый сервис реализует логику ТЗ
Добавлены новые эндпоинты для новой бизнес логики.

EvgenyAlexandrovichM committed:
11 спринт. 1 Ревью.
Исправлен ErrorHandler.java. Добавлен перехват серверных ошибок
Старый сервис переименован в storage, новый сервис реализует логику ТЗ
Добавлены новые эндпоинты для новой бизнес логики.
(Изменена структура JSON под тесты класса ErrorHandler.java)

EvgenyAlexandrovichM committed:
11 спринт. 1 Ревью.
Исправлен ErrorHandler.java. Добавлен перехват серверных ошибок
Старый сервис переименован в storage, новый сервис реализует логику ТЗ
Добавлены новые эндпоинты для новой бизнес логики.
(Изменена структура JSON под тесты класса ErrorHandler.java)

irek-valiev commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -17,26 +19,67 @@
 
+    private final FilmStorage filmStorage;
     private final FilmService filmService;
 
```
не обращайся к методам filmStorage напрямую из контроллера. Ты этим самым привязываешь контроллер не только к сервису, но и к DAO. И после замены в сервисе на класс работающий с БД, придется еще и контроллер править, а это не есть гуд. Задача контроллера знать у какого сервиса какой метод вызвать

irek-valiev commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -17,27 +19,77 @@
 
+    private final UserStorage userStorage;
     private final UserService userService;
 
```
аналогично про userStorage

irek-valiev commented on src/main/java/ru/yandex/practicum/filmorate/service/impl/FilmServiceImpl.java:
```diff
@@ -1,55 +1,74 @@
+        log.info("Получение топ 10 популярных фильмов по количеству лайков.");
+
+        return filmStorage.findAll().stream()
+                .sorted((f1, f2) -> f2.getLikes().size() - f1.getLikes().size())
```
это не обязательный комментарий, но я бы сортировку и limit перенес в InMemoryFilmStorage. Поясню причину. Пока хранилище мапа разницы нет где проводятся эти процедуры. Но после смены мапы на БД, придется точно переписывать, так как тащить все и только потом отсекать не нужное - ужасная практика. Все БД умеют сортировать и возвращать только нужное количество элементов. А это значит потом нужно будет править и сервисный слой. А разработку нужно вести так, чтобы замена одной реализации DAO-прослойки никак не влияла на сервисный слой

irek-valiev commented on src/main/java/ru/yandex/practicum/filmorate/service/impl/FilmServiceImpl.java:
```diff
@@ -1,55 +1,74 @@
+    private User getUserOrThrow(Long id) {
+        return userStorage.findUserById(id)
+                .orElseThrow(() -> new NotFoundException("Пользователь с userId " + id + " не найден."));
+    } /* дублированный метод из сервиса пользователей, который является приватным, но необходим здесь (найти вариант
```
такое дублирование вполне нормально, не стоит доводить до абсурда следования каким-то принципам, в частности DRY. В данном случае DRY выполняется в рамках одного класса и выглядит вполне аккуратно

EvgenyAlexandrovichM committed:
11 спринт. 2 Ревью.
Убрана зависимость в контроллерах от хранилищ FilmStorage.java и UserStorage.java
Дополнены интерфейсы сервисов FilService.java и UserService.java
Сортировка по популярности фильмов перенесена в хранилище

EvgenyAlexandrovichM committed:
11 спринт. 2 Ревью.
Убрана зависимость в контроллерах от хранилищ FilmStorage.java и UserStorage.java
Дополнены интерфейсы сервисов FilService.java и UserService.java
Сортировка по популярности фильмов перенесена в хранилище
